File: algorithms/fcfs/nim/fcfs_no_sort.nim
import strformat, sequtils

type
  Process = object
    pid: Natural
    at: Natural
    bt: Natural
    ct: Natural
  
var prc = @[Process(pid: 0, at: 5, bt: 3, ct: 0), Process(pid: 1, at: 2, bt: 1, ct: 0),
            Process(pid: 2, at: 0, bt: 1, ct: 0), Process(pid: 3, at: 4, bt: 2, ct: 0),
            Process(pid: 4, at: 0, bt: 2, ct: 0), Process(pid: 5, at: 2, bt: 3, ct: 0)]
var ta = 0.0
var w = 0.0

proc `<`(x, y: Process): bool =
  if x.at < y.at:
    true
  else:
    false

proc findCompletionTime(): void =
  var fin = newSeq[bool](len(prc))
  var time = 0
  var idx = 0
  var over = 0
  var lastArrival = max(prc).at + 1
  while over < len(prc):
    var curr = lastArrival
    for p in zip(fin, prc):
      if (not p[0] and p[1].at < curr):
        curr = p[1].at
        idx = p[1].pid
    if (time < prc[idx].at):
      time = prc[idx].at
    fin[idx] = true
    time += prc[idx].bt
    prc[idx].ct = time
    over += 1

findCompletionTime()

echo "PID\tAT\tBT\tCT\tTAT\tWT"
for p in prc:
  let pTA = (float)(p.ct - p.at)
  let pW = pTA - (float)(p.bt)
  echo &"{p.pid}\t{p.at}\t{p.bt}\t{p.ct}\t{(int)pTA}\t{(int)pW}"
  ta += pTA / (float)(len(prc))
  w += pW / (float)(len(prc))

echo &"Avg. TAT: {ta:.3f}"
echo &"Avg. WT: {w:.3f}"

File: algorithms/fcfs/fcfs-no-sort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

class Process {
    int pid, at, bt, ct;

    public Process(int pid, int at, int bt) {
        this.pid = pid;
        this.at = at;
        this.bt = bt;
        this.ct = 0;
    }
}

class fcfs {
    public static void main(String[] args) {
        ArrayList<Process> proc = new ArrayList<Process>(
                Arrays.asList(new Process(0, 1, 5), new Process(1, 0, 5), new Process(2, 0, 2), new Process(3, 2, 4)));
        float ta = 0, w = 0;
        ArrayList<Boolean> fin = new ArrayList<Boolean>(Collections.nCopies(proc.size(), false));

        int max = Collections.max(proc, (Process a, Process b) -> {
            if (a.at < b.at)
                return -1;
            return 1;
        }).at + 1;
        int time = 0, end = 0;

        while (end < proc.size()) {
            int idx = 0, min = max;
            for (int i = 0; i < proc.size(); i++)
                if (!fin.get(i) && proc.get(i).at < min) {
                    min = proc.get(i).at;
                    idx = i;
                }
            if (time < proc.get(idx).at)
                time = proc.get(idx).at;
            fin.set(idx, true);
            time += proc.get(idx).bt;
            proc.get(idx).ct = time;
            end++;
            ta += proc.get(idx).ct - proc.get(idx).at;
            w += proc.get(idx).ct - (proc.get(idx).at + proc.get(idx).bt);
        }

        System.out.println("PID\tAT\tBT\tCT\tTAT\tWT");
        for (Process p : proc) {
            System.out.println(p.pid + "\t" + p.at + '\t' + p.bt + '\t' + p.ct + '\t' + (p.ct - p.at) + '\t'
                    + (p.ct - (p.at + p.bt)));
        }
        System.out.println("Avg. TAT: " + ta / proc.size());
        System.out.println("Avg. WT: " + w / proc.size());
    }
}